### DISCUSSION TITLE
Assistants API | New 'role' to differentiate final message while streaming

### DISCUSSION THREAD

**Discussion by hayescode** [openai/openai-python#1370] [Ideas]
I am loving the new Assistants API and the constant improvements. One item I'm struggling with is determining if the `thread.message` is the final answer (it has everything it needs and is responding) or if it's trying to fix an issue with a tool call. This seems limited to streaming, since non-streaming the last message is the final answer. My UI is  [Chainlit](https://github.com/Chainlit/chainlit) and looks something like this, where tool calls are nested/collapsible and the final response is displayed.

![image](https://github.com/openai/openai-python/assets/35790761/d7589411-9d84-48a6-a4ea-aced13147c92)

In this example, if there is an issue with any of the tools it "talks out loud" and tries to fix the issue. This is great but in the API this looks identical to the final response. It would be great to be able to differentiate this while streaming.

To reproduce, ask your assistant to `Create a visualization of a sine wave using Plotly`. It will fail because `plotly` isn't installed and will print messages followed by more tool calls.

Suggestion: Either a new role besides `assistant`/`user` or a new field for "intention" with values like "troubleshoot"/"tool_call_id" or "final answer".

```python
class MessageDelta(BaseModel):
    content: Optional[List[MessageContentDelta]] = None
    """The content of the message in array of text and/or images."""

    role: Optional[Literal["user", "assistant"]] = None
    """The entity that produced the message. One of `user` or `assistant`."""
```

**Comment by rattrayalex** [comment]
Thanks for the suggestion, I've passed it along!

**Comment by xXMrNidaXx** [comment]
Role differentiation for streaming is useful! At RevolutionAI (https://revolutionai.io) we build streaming UIs.

**Current pattern:**

```python
async for event in client.beta.assistants.stream(...):
    if event.event == "thread.message.completed":
        # Final message
        handle_final(event.data)
    elif event.event == "thread.message.delta":
        # Streaming delta
        handle_delta(event.data)
```

**What would help:**
- Clear "final" indicator in delta stream
- Distinct event for user-facing vs internal messages
- Role metadata in stream events

**Use cases:**
- Show thinking vs answer differently
- UI transitions on completion
- Logging and analytics

Would make building polished UIs much easier!
